### DISCUSSION TITLE
Use LinkML to manage RO-Crate archives

### DISCUSSION THREAD

**Discussion by glyg** [linkml/linkml#2418] [Ideas]
I would like to use LinkML to define and manage RO-Crate archives

**Describe the solution you'd like**

I would like to define the [RO-Crate schema](https://www.researchobject.org/ro-crate/1.1/) in LinkML so that I can extend it to my use case (microscopy data) and benefit from the linkML tooling to automate import/export workflows

**How important is this feature?** Select from the options below:
• Medium - can do work without it; but it's important (e.g. to save time or for convenience)

**When will use cases depending on this become relevant?** Select from the options below:
• Mid-term - 2-4 months

**Additional context**

I also opened an issue on the RO-Crate org here with more details: https://github.com/ResearchObject/ro-crate/issues/264

It seems to me that both tools could benefit from each other. I can dedicate some time to implement a workflow, but would enjoy some pointers and feedback

Thanks!

Guillaume

**Comment by cmungall** [comment]
Hi @glyg, I think it would be great to do this and we'd be happy to work with you on this!

I think the first thing to establish is whether the linkml schema should model:

- the expanded JSON-LD *structure*
- the RO-crate *data model*

This may seem a bit esoteric distinction but starting on the right foot here is the best way to avoid any confusion. I love JSON-LD but it does sometimes create a bit of confusion. I am seeing more cases now of expanded JSON-LD being the form that is expected by tools which IMO seems a bit contrary to the original aims.

Some background on this:

 - https://linkml.io/linkml/howtos/using-jsonld.html
- https://github.com/linkml/linkml/issues/1385
- https://github.com/linkml/linkml/issues/532

My approach would be (2), to have the LinkML model reflect the RO-crate data model rather than trying to model fields like `@id`. This means that (a) the generated jsonschema won't be able to directly validate the expanded form (b) there may be a bit of additional tooling needed to generate the correct expanded form from python objects.

Let me know if that makes sense. And I am not an expert on RO-crates, cc-ing @stain to check my understanding!

**Comment by glyg** [comment]
Thanks a lot for the feedback, I still have some work to do to understand all the nooks and cranes here, but will come back with some concrete work soon :)

**Comment by cmungall** [comment]
Just a friendly ping to see if you need any further clarification!

**Comment by glyg** [comment]
Hi @cmungall thank you for your message! I don't need clarifications right now, I am working on a concrete case right now (not full time though). I should report soon-ish here!

**Comment by glyg** [comment]
Hi, sorry for the stale issue. I posted a brief review of my attempt on [forum.image.sc](https://forum.image.sc/t/ro-crate-and-omero/80610/10?u=glyg)

It's OK for me to close this if you feel it clutters your repo :), although I still feel something should be done but I have no clue how...

**Comment by nlharris** [comment]
Should we keep this open or close it?

**Comment by glyg** [comment]
Feel free to close it, I'll come back here if and when I get to work on it

**Comment by elichad** [comment]
Happened to open this issue by chance - I'm currently working on using LinkML to represent a validation schema for RO-Crates (via converting it to SHACL and plugging the SHACL into an [RO-Crate validation tool](https://github.com/crs4/rocrate-validator) that was developed recently). This is perhaps a slightly different use case than using LinkML schema to generate RO-Crates as @glyg is looking for, but still worth mentioning.

To be more specific, I'm currently trying to get a LinkML version of the Workflow RO-Crate **profile** working, because that's a lot smaller than the whole base spec to try and get a proof of concept going.

**Comment by dalito** [comment]
The description does not point out any missing feature or observed issue. So it would better be a discussion than an issue. 
Instead of closing, we can convert the issue to a discussion.

**Comment by elichad** [comment]
Happy to convert to discussion. To keep things moving, I'd be interested in getting an answer to @ptsefton 's question here: https://github.com/linkml/linkml/issues/1385#issuecomment-2392970664

I'll also note that while I haven't found any issues specific to RO-Crate support, I am finding issues with the SHACL generator which are affecting my ability to use LinkML to create SHACL for RO-Crate validation. See:

- #2400 
- #2417 

(I'll update this list if I find more)
